Fix leaks from reference-counted types in return position

toasteater opened the pull request:
`GodotObject::from_sys` has Clone semantics on reference-counted types, which is relied on by macros. However, any references returned from ptrcalls are leaked in the process of being cast into a pointer. This caused an extra add_ref to be called on references in return position, leading to memory leaks.

This adds an extra `from_return_position_sys` method on `GodotObject`, which doesn't call `add_ref` for reference-counted types, which is used for ptrcalls in the generated bindings, fixing the leak.

toasteater commented:
I'll think about how this can be tested automatically. We can only get drop callbacks through a NativeScript, so maybe we find a property that takes a reference-counted type, and do:

1. Make instance of probe NativeScript.
2. Set property to instance.
3. Drop instance.
4. Get reference using ptrcall.
5. Set property to another value, make sure we have the only reference.
6. Drop reference.
7. Assert that the NativeScript is dropped.

Does this look like a sound plan @karroffel?

karroffel commented:
That sounds like a good plan to me. It will be fun finding a native type like that, but I am sure there is something like that.

toasteater commented:
Added the reference leak test, and split the tests crate into several modules because it's getting large.

I put some comments in the test, since what it's doing is probably not the most obvious. I'll remove some of them if it's excessive.

karroffel commented:
It seems good to me, no worries about too many comments :smile:
